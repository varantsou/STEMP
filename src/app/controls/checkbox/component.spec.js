import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Checkbox from './component';

const propsWithClassName = {
    className: 'test-class'
};

const propsWithChecked = {
    checked: true
};

const propsWithDisabled = {
    disabled: true
};

const propsWithValidate = {
    validate: true
};

const propsWithOnChanged = {
    onChanged: jest.fn()
};

const propsRequired = {
};

const propsDefault = {
    ...propsRequired,
    ...propsWithClassName,
    ...propsWithChecked,
    ...propsWithDisabled,
    ...propsWithValidate,
    ...propsWithOnChanged
};

describe('Should render Checkbox component', () => {
    test('with default properties', () => {
        const component = shallow(<Checkbox {...propsDefault} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });

    test('with required properties', () => {
        const component = shallow(<Checkbox {...propsRequired} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });

    test('with "className" property', () => {
        const component = shallow(<Checkbox {...propsWithClassName} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });

    test('with "checked" property', () => {
        const component = shallow(<Checkbox {...propsWithChecked} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });

    test('with "disabled" property', () => {
        const component = shallow(<Checkbox {...propsWithDisabled} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });

    test('with "validate" property', () => {
        const component = shallow(<Checkbox {...propsWithValidate} />);
        const output = shallowToJson(component);

        expect(output).toMatchSnapshot();
    });
});

describe('Checkbox component', () => {
    test('should implement "renderValue" method when state.checked is true', () => {
        const component = shallow(<Checkbox {...propsRequired} />);
        const instance = component.instance();

        instance.state = {
            checked: true
        };

        const classNames = {
            value: 'test-value',
            valueIcon: 'test-value-icons'
        };

        const output = instance.renderValue(classNames);

        expect(output).toMatchSnapshot();
    });

    test('should implement "handleClick" method', () => {
        const component = shallow(<Checkbox {...propsRequired} />);
        const instance = component.instance();

        jest.spyOn(instance, 'toggleChecked')
            .mockImplementation(() => {});

        instance.handleClick();

        expect(instance.toggleChecked).toBeCalled();
    });

    describe('Should implement "handleKeyDown" method', () => {
        const event = {
            preventDefault: jest.fn()
        };

        test('if event.keyCode is equal to 32', () => {
            const component = shallow(<Checkbox {...propsRequired} />);
            const instance = component.instance();

            jest.spyOn(instance, 'toggleChecked')
                .mockImplementation(() => {});

            event.keyCode = 32;

            instance.handleKeyDown(event);

            expect(event.preventDefault).toBeCalled();
            expect(instance.toggleChecked).toBeCalled();
        });

        test('if event.keyCode is NOT equal to 32', () => {
            const component = shallow(<Checkbox {...propsRequired} />);
            const instance = component.instance();

            event.keyCode = 31;

            instance.handleKeyDown(event);
        });
    });

    describe('Should implement "toggleChecked" method', () => {
        test('without "onChanged" property', () => {
            const component = shallow(<Checkbox {...propsRequired} />);
            const instance = component.instance();

            jest.spyOn(instance, 'setState')
                .mockImplementation(() => {});

            instance.state = {
                checked: false
            };

            instance.toggleChecked();

            expect(instance.setState).toBeCalledWith({ checked: true });
        });

        test('with "onChanged" property', () => {
            const component = shallow(<Checkbox {...propsWithOnChanged} />);
            const instance = component.instance();

            jest.spyOn(instance, 'setState')
                .mockImplementation(() => {});

            jest.spyOn(instance.props, 'onChanged')
                .mockImplementation(() => {});

            instance.state = {
                checked: true
            };

            instance.toggleChecked();

            expect(instance.setState).toBeCalledWith({ checked: false });
            expect(instance.props.onChanged).toBeCalledWith(false);
        });
    });

    describe('Should implement "componentDidUpdate" method', () => {
        test('with change "checked" property', () => {
            const checked = true;
            const testProps = {
                checked
            };
            const component = shallow(<Checkbox {...testProps} />);
            const instance = component.instance();
            const oldChecked = false;
            const prevProps = {
                checked: oldChecked
            };

            jest.spyOn(instance, 'setState')
                .mockImplementation(() => {});

            instance.componentDidUpdate(prevProps);

            expect(instance.setState).toBeCalledWith({ checked: checked });
        });

        test('without change "checked" property', () => {
            const checked = true;
            const testProps = {
                checked
            };
            const component = shallow(<Checkbox {...testProps} />);
            const instance = component.instance();
            const oldChecked = true;
            const prevProps = {
                checked: oldChecked
            };

            jest.spyOn(instance, 'setState')
                .mockImplementation(() => {});

            instance.componentDidUpdate(prevProps);

            expect(instance.setState).not.toBeCalled();
        });
    });
});
