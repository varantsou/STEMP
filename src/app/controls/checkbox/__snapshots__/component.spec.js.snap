// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should implement "renderValue" method when state.checked is true 1`] = `
<div
  className="test-value"
>
  <img
    alt=""
    className="test-value-icon"
    src="test-file-stub"
  />
</div>
`;

exports[`Should render Checkbox component with "checked" property 1`] = `
<div
  aria-checked={true}
  className="checkbox checkbox--checked"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  >
    <img
      alt=""
      className="checkbox__value-icon"
      src="test-file-stub"
    />
  </div>
</div>
`;

exports[`Should render Checkbox component with "className" property 1`] = `
<div
  aria-checked={false}
  className="checkbox test-class"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  />
</div>
`;

exports[`Should render Checkbox component with "disabled" property 1`] = `
<div
  aria-checked={false}
  aria-disabled={true}
  className="checkbox checkbox--disabled"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  />
</div>
`;

exports[`Should render Checkbox component with "validate" property 1`] = `
<div
  aria-checked={false}
  className="checkbox"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  />
</div>
`;

exports[`Should render Checkbox component with default properties 1`] = `
<div
  aria-checked={true}
  aria-disabled={true}
  className="checkbox test-class checkbox--checked checkbox--disabled"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  >
    <img
      alt=""
      className="checkbox__value-icon"
      src="test-file-stub"
    />
  </div>
</div>
`;

exports[`Should render Checkbox component with required properties 1`] = `
<div
  aria-checked={false}
  className="checkbox"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex="0"
>
  <div
    className="checkbox__value"
  />
</div>
`;
